attrs_strict/_type_validation.py

```diff
@@ -91,12 +98,31 @@ def _get_base_type(type_):
 
 
 def _type_matching(actual, expected):
-    return expected == actual or (
-        _get_base_type(expected) == typing.Union
-        and any(
-            candidate_type == actual for candidate_type in expected.__args__
-        )
-    )
+    if expected == actual or expected == typing.Any:
+        return True
+
+    base_type = _get_base_type(expected)
+
+    if base_type == typing.Union:
+        if any(
+            _type_matching(actual, expected_candidate)
+            for expected_candidate in expected.__args__
+        ):
+            return True
+
+    elif base_type in (
+        SimilarTypes.Dict
+        | SimilarTypes.List
+        | SimilarTypes.Tuple
+        | SimilarTypes.Callable
+    ):
+        if all(
+            _type_matching(a, e)
+            for a, e in zip_longest(actual.__args__, expected.__args__)
```

Comment:
consider actual, expected instead of `a` and `e` for clarity, it might not be that clear for non-native speakers.